File: LeetCode题目集/125. Valid Palindrome(easy).java
/*
 * @Author: SourDumplings
 * @Link: https://github.com/SourDumplings/
 * @Email: changzheng300@foxmail.com
 * @Description: https://leetcode.com/problems/valid-palindrome/
 * @Date: 2019-03-20 18:43:18
 */

class Solution
{
    public boolean isPalindrome(String s)
    {
        StringBuffer str = new StringBuffer();
        int l = s.length();
        for (int i = 0; i < l; i++)
        {
            char c = s.charAt(i);
            if (Character.isAlphabetic(c) || Character.isDigit(c))
            {
                str.append(Character.toLowerCase(c));
            }
        }
        int ll = str.length();
        for (int i = 0; i < ll / 2; ++i)
        {
            if (str.charAt(i) != str.charAt(ll - i - 1))
            {
                return false;
            }
        }
        return true;
    }
}

File: OJ practices/LeetCode题目集/125. Valid Palindrome(easy).cpp
/**
 * @file 125. Valid Palindrome(easy).cpp
 * @author chadchang (chadchang@tencent.com)
 * @brief 
 * @version 1.0.0
 * @date 2024-03-28
 * 
 * @copyright Copyright (c) 2024 SourDumplings
 * 
 * https://leetcode.cn/problems/valid-palindrome/description/?envType=study-plan-v2&envId=top-interview-150
 */

class Solution
{
public:
    bool isPalindrome(string s)
    {
        int n = s.length();
        int l = 0, r = n - 1;
        while (l < n)
        {
            while (l < n && !isalnum(s[l])) ++l;
            while (l <= r && !isalnum(s[r])) --r;
            if (r <= l)
            {
                break;
            }
            
            char lc = tolower(s[l]);
            char rc = tolower(s[r]);
            if (lc != rc)
            {
                return false;
            }
            ++l;
            --r;
        }
        return true;
    }
};